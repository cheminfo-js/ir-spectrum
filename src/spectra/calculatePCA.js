import PCA from 'ml-pca';

export function calculatePCA(spectra) {
  if (spectra.cache.pca) return spectra.cache.pca;
  let matrix = spectra.spectra.map((entry) => entry.normalized);

  const pca = new PCA(matrix);
  pca.matrix = matrix;
  spectra.cache.pca = pca;
  return pca;
}
