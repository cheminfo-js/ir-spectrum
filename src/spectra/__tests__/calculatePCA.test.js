import { calculatePCA } from '../calculatePCA';

test('getPCA', () => {
  let spectra = {
    cache: {},
    spectra: [
      { normalized: [1, 1, 1, 1, 1, 1, 1, 1] },
      { normalized: [1, 0, 1, 1, 1, 1, 1, 1] },
      { normalized: [1, 1, 1, 1, 1, 1, 1, 1] },
      { normalized: [1, 0, 1, 1, 0, 1, 0, 1] },
      { normalized: [1, 1, 1, 1, 0, 1, 0, 1] },
      { normalized: [1, 0, 1, 1, 0, 1, 0, 1] }
    ]
  };

  let pca = calculatePCA(spectra);

  let result = pca.predict(pca.matrix, { nComponents: 2 });
  expect(result).toHaveLength(6);
  expect(result[0]).toHaveLength(2);
  expect(result).toMatchSnapshot();
});
