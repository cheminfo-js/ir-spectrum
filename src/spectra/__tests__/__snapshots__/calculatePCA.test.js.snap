// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPCA 1`] = `
Matrix [
  Array [
    -0.8417163919188921,
    0.20374865784844376,
  ],
  Array [
    -0.4726682074693537,
    -0.7256616054661487,
  ],
  Array [
    -0.8417163919188921,
    0.20374865784844376,
  ],
  Array [
    0.8417163919188921,
    -0.20374865784844376,
  ],
  Array [
    0.4726682074693537,
    0.7256616054661487,
  ],
  Array [
    0.8417163919188921,
    -0.20374865784844376,
  ],
]
`;
